compiler/rustc_resolve/src/access_levels.rs

```diff
@@ -0,0 +1,128 @@
+                } else {
+                    None
+                }
+            }
+        };
+
+        let access_level = self.r.set_access_level(item.id, inherited_item_level);
+
+        match item.kind {
+            ast::ItemKind::ForeignMod(ForeignMod { ref items, .. }) => {
+                for nested in items {
+                    if nested.vis.kind.is_pub() {
+                        self.r.set_access_level(nested.id, access_level);
+                    }
+                }
+            }
+            ast::ItemKind::Enum(EnumDef { ref variants }, _) => {
+                for variant in variants {
+                    let variant_level = self.r.set_access_level(variant.id, access_level);
+                    if let Some(ctor_id) = variant.data.ctor_id() {
+                        self.r.set_access_level(ctor_id, access_level);
+                    }
+
+                    for field in variant.data.fields() {
+                        self.r.set_access_level(field.id, variant_level);
+                    }
+                }
+            }
+            ast::ItemKind::Struct(ref def, _) | ast::ItemKind::Union(ref def, _) => {
+                if let Some(ctor_id) = def.ctor_id() {
+                    self.r.set_access_level(ctor_id, access_level);
+                }
+
+                for field in def.fields() {
+                    if field.vis.kind.is_pub() {
+                        self.r.set_access_level(field.id, access_level);
+                    }
+                }
+            }
+            ast::ItemKind::Trait(ref trait_kind) => {
+                for nested in trait_kind.4.iter() {
+                    self.r.set_access_level(nested.id, access_level);
+                }
+            }
+            ast::ItemKind::Impl(ref impl_kind) => {
+                for nested in impl_kind.items.iter() {
+                    if impl_kind.of_trait.is_some() || nested.vis.kind.is_pub() {
+                        self.r.set_access_level(nested.id, access_level);
+                    }
+                }
+            }
+            ast::ItemKind::Mod(..) => {
+                if access_level.is_some() {
+                    self.r.set_exports_access_level(self.r.local_def_id(item.id));
+                }
+            }
+            ast::ItemKind::ExternCrate(..)
+            | ast::ItemKind::Use(..)
+            | ast::ItemKind::Static(..)
+            | ast::ItemKind::Const(..)
+            | ast::ItemKind::GlobalAsm(..)
+            | ast::ItemKind::TyAlias(..)
+            | ast::ItemKind::TraitAlias(..)
+            | ast::ItemKind::MacroDef(..)
+            | ast::ItemKind::MacCall(..)
+            | ast::ItemKind::Fn(..) => {
+                return;
+            }
+        }
+
+        let orig_level = std::mem::replace(&mut self.prev_level, access_level);
+        visit::walk_item(self, item);
+        self.prev_level = orig_level;
+    }
+
+    fn visit_block(&mut self, block: &'ast ast::Block) {
```

Comment:
In this visitor we can avoid visiting blocks entirely.
To detect public and exported items we only need to recurse into `ItemKind::Mod` items, in all other cases we can return without calling `visit::walk_item`.